### PR TITLE
[Core] Support branch batch/streaming read and write

### PR DESCRIPTION
**Pull request by TaoZex**
### Purpose
Write: After data is written, the new snapshot is written to the corresponding branch directory
Read: Read the snapshot in the corresponding branch directory to get the data file position

Linked issue: 
https://github.com/apache/incubator-paimon/issues/2727
https://github.com/apache/incubator-paimon/issues/2726

### Next

support merge branch

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:39]: What's this `MAIN_BRANCH` used for?

Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:39]: Method changes are similar to the following figure:
image
By default, we treat it as the main branch. For compatibility with the previous design, branchName = "main" is called if the method is called with no arguments.

File: paimon-core/src/main/java/org/apache/paimon/AppendOnlyFileStore.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/AppendOnlyFileStore.java:72]: Use MAIN_BRANCH instead of "main"

File: paimon-core/src/main/java/org/apache/paimon/AppendOnlyFileStore.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/AppendOnlyFileStore.java:106]: Use MAIN_BRANCH instead of "main"

File: paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java:110]: Use MAIN_BRANCH instead of "main"

File: paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java:166]: Use MAIN_BRANCH instead of "main"

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:715]: We can compare branch name and get path for table in BranchManager.getBranchPath instead of here

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (FangYongs) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:39]: Rename the field name to DEFAULT_MAIN_BRANCH and if paimon uses "main" as default main, you need to add validation for branch name when users create branches

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:715]: It will call the getBranchPath method.
image

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java]: Thanks for your review. I have modified the code according to your suggestion, please review it again.

File: paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java
Comment (FangYongs) [inline paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:281]: We should not use `System.out.pringln` in tests

Comment (TaoZex) [reply paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java]: Thanks for your reminder. Done. @FangYongs

File: paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java
Comment (xiangyuf) [inline paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java:65]: Why not use `BranchManager.DEFAULT_MAIN_BRANCH` directly here.

Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/AbstractFileStore.java:65]: I have fixed it. Thanks. @xiangyuf

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (xiangyuf) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:79]: Use `DEFAULT_MAIN_BRANCH ` instead of 'main' here.

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (xiangyuf) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:81]: `tagManager.tagExists` only checks the table path and tag path. I think we should also check the branch path here.

**Review by xiangyuf — CHANGES_REQUESTED** [review]

**Comment by xiangyuf** [comment]
Hi @TaoZex, I browsed through your request and thx for ur work. In my humble opinion, this PR contains lots of changes and can be separated into several PRs in order to improve the efficiency of code review. For example, we can have one PR for changes in `BranchManager`, `TagManager`,  `SnapShotManager` and `SchemaManager` to support some basic operations of Branch. After that, we can launch another PR for branch Streaming/Batch read or Streaming/Batch write. By doing this, we can make the code review work more easy here. And once the basic operations of Branch have been merged, we can implement the replace/merge operation of Branch. WDYT? Also, I'd like to hear more from @FangYongs .

File: paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java
Comment (FangYongs) [inline paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java:1264]: We also need to verify that the data of the `main` is not affected by `branch1`

Comment (TaoZex) [reply paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java:1272]: I've already added it, thanks for the suggestion. Please open ci.

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:81]: I have fixed it.

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:83]: The branchExists method already checks the branch path.

**Comment by TaoZex** [comment]
> Hi @TaoZex, I browsed through your request and thx for ur work. In my humble opinion, this PR contains lots of changes and can be separated into several PRs in order to improve the efficiency of code review. For example, we can have one PR for changes in `BranchManager`, `TagManager`, `SnapShotManager` and `SchemaManager` to support some basic operations of Branch. After that, we can launch another PR for branch Streaming/Batch read or Streaming/Batch write. By doing this, we can make the code review work more easy here. And once the basic operations of Branch have been merged, we can implement the replace/merge operation of Branch. WDYT? Also, I'd like to hear more from @FangYongs .

Thank you for your suggestion. We have discussed related issues in https://github.com/apache/incubator-paimon/pull/2608. In order to be compatible with the original design and reduce changes, the form of changes is similar to the following figure:
image
The current approach is the result of the previous discussion, and there are not many changes shown in the figure above, so I still hope to merge the current pr. Thanks again for your advice.

**Comment by TaoZex** [comment]
Timeout in ci, please rerun it. Thanks.

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:423]: typo: branName -> branchName

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:383]: This piece of code appeared many times, could you provide a method to express this directory logic simply?

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: This is not implemented completely right? Because there isn't any other caller for this method and pass a `branchName` besides `DEFAULT_MAIN_BRANCH`. Are you planning to implement this in the next pr?

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java:44]: What's this property used for?

File: paimon-core/src/main/java/org/apache/paimon/schema/TableSchema.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/schema/TableSchema.java:281]: What's this used for?

File: paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java
Comment (schnappi17) [inline paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java:73]: Should this be used when counting tags or filter tags to distinguish the tags in default main branch and custom branches?

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:383]: Good idea！ I will fix it.

File: paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java
Comment (schnappi17) [inline paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java:1513]: What's the difference with get `branchManager` by `table.branchManager()` directly?

File: paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java
Comment (schnappi17) [inline paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:89]: Could we also verify the main branch is not affected?

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: In AbstractFileStoreTable line 320, this method is invoked.

Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: Example : In FileStoreTableTestBase line 1284

File: paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java
Comment (schnappi17) [reply paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:89]: Ok, see it's verified in the `FileStoreTableTestBase`, I think it's enough. Please ignore this comment.

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java
Comment (schnappi17) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: I mean if we've created a branch for Paimon table through procedure, how to really write to this custom branch in a Flink job? It seems like the committer operator in Flink job is only able to write to the default branch currently.

Comment (schnappi17) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: Maybe there should be a option for Flink connector in `FlinkConnectorOptions` to set a writing branch for Flink jobs, and in Flink sink this option could be recognized to create a branched commit. But I think these could be implemented in the next PR for the scope for this PR is big enough, breaking code to PRs can make the code more clear to track and review.

File: paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java:73]: Yes, assuming the current table directory is table, then the tags of default main branch are in table/tag, and the tags of custom branches are in table/custom_branch/tag. The branch to which the tag belongs depends on the location.

File: paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:302]: I agree with your suggestion, and I plan to support flink and spark read and write after this pr merge. As you suggest, this is more conducive to review.

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java]: Thanks. Done.

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java]: Removed.

File: paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java
Comment (TaoZex) [reply paimon-core/src/test/java/org/apache/paimon/table/FileStoreTableTestBase.java]: No difference. Has been replaced by `table.branchManager()`.

File: paimon-core/src/main/java/org/apache/paimon/schema/TableSchema.java
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/schema/TableSchema.java]: It is used in merge branch, so I temporarily deleted it.

**Comment by schnappi17** [comment]
+1 for this, thanks @TaoZex for contribution

**Comment by TaoZex** [comment]
Thanks for every reviewer. Happy New Year! @FangYongs @schnappi17 @xiangyuf

**Comment by xiangyuf** [comment]
@TaoZex Great work, thx for contribution.

**Pull request merged**
